rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/internal/LRARecoveryModule.java

```diff
@@ -130,6 +135,40 @@ private void doRecoverTransaction(Uid recoverUid) {
         }
     }
 
+    private void moveEntryToFailedLRAPath (final Uid failedUid) throws ObjectStoreException {
+        String failedLRAType = LongRunningAction.getFailedLraType();
+        InputObjectState inputState = _recoveryStore.read_committed(failedUid, _transactionType);
+        InputObjectState failedLRAUidState = _recoveryStore.read_committed(failedUid, failedLRAType);
+        boolean result = false;
+
+        if (inputState != null) {
+            if (failedLRAUidState != null) {
+                // Record already exists in failedLRARecord location, hence removing it from the LRARecord location
+                boolean rmRec = _recoveryStore.remove_committed(failedUid, _transactionType);
+                if (!rmRec) {
+                    LRALogger.i18NLogger.warn_UnableToRemoveDuplicateFailedLRARecord(
+                            failedUid.toString(), failedLRAType, _transactionType);
+                    return;
+                }
+                LRALogger.logger.infof("The failed lra record (Uid: %s) is already present in the failed LRA record location type: %s." +
+                        "Hence removed the failed duclicate record from LRARecord location: %s", failedUid, failedLRAType, _transactionType);
+                return;
```

Comment:
why two return statements if (!rmRec). Easy to read would be:

```java
                if (!_recoveryStore.remove_committed(failedUid, _transactionType)) {
                    LRALogger.logger.infof("The failed lra record (Uid: %s) is already present in the failed LRA record location type: %s." +
                            "Hence removed the failed duclicate record from LRARecord location: %s", failedUid, failedLRAType, _transactionType);
                } else {
                    LRALogger.i18NLogger.warn_UnableToRemoveDuplicateFailedLRARecord(
                            failedUid.toString(), failedLRAType, _transactionType);
                }
                return;
```
Under what circumstances will the record exist in two parts of the store (should only happen if the server has previously crashed or are there other scenarios).

If the delete failed then the warning will be repeated forever after.

Why is this logged at warning level, what are the consequences of being unable to delete the record. Surely it is more informational but if you want to log at warn level then only do it once.

Reply:
Also there is too much logging going on here, moving a record can now produce four different log statements. Why is the existing ArjunaCore code insufficient:

```java
	    InputObjectState state = _recoveryStore.read_committed(newUid, _typeName);
	    boolean res = false;
	    
	    if (state != null) // just in case recovery
	        // kicked-in
	    {
	        boolean moved = _recoveryStore.write_committed(newUid, _movedTypeName, new OutputObjectState(state));

	        if (!moved) {
                tsLogger.logger.debugf("Removing old transaction status manager item %s", newUid);
                }
	        else {
	            res = _recoveryStore.remove_committed(newUid, _typeName);
                    tsLogger.i18NLogger.warn_recovery_ExpiredTransactionStatusManagerScanner_6(newUid);
                }

	    }
```

Reply:
Okay, I will log warn level only once.

Here is the scenario in which the record will be available in both the location:
1. Failed records are in `/StateManager/BasicAction/LongRunningAction/` -> RecoveryManager Kicked in -> reading from `/StateManager/BasicAction/LongRunningAction/` for first time -> writing the failed record at `/StateManager/BasicAction/LongRunningAction/Failed` -> now deleting/removing the failed record from `/StateManager/BasicAction/LongRunningAction/` -> failed to `delete/remove` due to system crash or maybe for some other reason.

Now at this point of time we will have the failed record in both the locations.

Now recoveryManager kicks in for second time -> from here we can check if the record already exists in `/StateManager/BasicAction/LongRunningAction/Failed`?

Reply:
Okay